### PR TITLE
WindowsAppRuntime_MSIXInstallFromPath.exe

### PR DESCRIPTION
**Pull request by DrusTheAxe**
Helper tool to register packages on F5

NOTE: This just creates the exe and adds it to the nuget. VS logic to invoke it will be in a follow up PR

Tested manually via command line invocation e.g.

Step 1. Create MSIX.inventory (like where the Aggregator pipeline now does) in the NuGet package e.g.

```
Microsoft.WindowsAppRuntime.1.0.msix=Microsoft.WindowsAppRuntime.1.0_0.258.2323.0_x64__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.Main.1.0.msix=Microsoft.WindowsAppRuntime.Main.1.0_0.258.2323.0_x64__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.Singleton.1.0.msix=Microsoft.WindowsAppRuntime.Singleton_0.258.2323.0_x64__8wekyb3d8bbwe
Microsoft.WindowsAppRuntime.DDLM.1.0.msix=Microsoft.WinAppRuntime.DDLM.0.258.2323.0-x6_0.258.2323.0_x64__8wekyb3d8bbwe
```

Step 2. Run WindowsAppRuntime_MSIXInstallFromPath.exe like VS will (in the upcoming PR)

```
WindowsAppRuntime_MSIXInstallFromPath.exe C:\source\repos\projectreunionInternal\packages\microsoft.windowsappsdk\1.0.0-20210916.19-ci\tools\MSIX\win10-x64
```

NOTE: There's some debuggery printf still in place. Will be removed in follow up PR that adds the VS invocation.

### CODE REVIEW

File: dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+#include <fstream>
+
+void Help();
+HRESULT JustDoIt(PCWSTR path) noexcept;
+void AddPackageIfNecessary(PCWSTR path, const std::wstring& filename, const std::wstring& packageFullName);
+bool NeedToRegisterPackage(const std::wstring& packageFullName);
+bool NeedToRegisterPackage(PCWSTR packageFullName);
+HRESULT AddPackage(PCWSTR path, const std::wstring& filename);
+
+class PackageId
```
Comment (DrusTheAxe) [inline dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp:16]: Will get moved to dev\common\AppModel.Package.h in a later PR. Needs some work before it's ready for general use so I didn't want to hold up this work on that cleanup/refactoring

File: dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        AddPackageIfNecessary(path, filename, packageFullName);
+    }
+
+    return 0;
+}
+CATCH_RETURN();
+
+void AddPackageIfNecessary(PCWSTR path, const std::wstring& filename, const std::wstring& packageFullName)
+{
+    wprintf(L"Path: %s\n", path);
```
Comment (DrusTheAxe) [inline dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp:173]: Debugging printfer will go away in the follow up PR after VS invoking this is tested

File: dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+            L"  path = Path where Windows App SDK's MSIX packages can be found\n"
+            L"           => Microsoft.WindowsAppRuntime.*.msix\n"
+            L"           => Microsoft.WindowsAppRuntime.DDLM.*.msix\n"
+            L"           => Microsoft.WindowsAppRuntime.Main.*.msix\n"
+            L"           => Microsoft.WindowsAppRuntime.Singleton.*.msix\n");
+}
+
+HRESULT JustDoIt(PCWSTR path) noexcept try
+{
+    wprintf(L"path: %s\n", path);
```
Comment (DrusTheAxe) [inline dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp:134]: Debugging printfer will go away in the follow up PR after VS invoking this is tested

File: dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+{
+    return NeedToRegisterPackage(packageFullName.c_str());
+}
+
+bool NeedToRegisterPackage(PCWSTR packageFullName)
+{
+    auto packageId{ PackageId::FromPackageFullName(packageFullName) };
+
+    // Is there an equal-or-better package already registered?
+    wprintf(L"Family: %s\n", packageId.PackageFamilyName());
```
Comment (DrusTheAxe) [inline dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp:200]: Debugging printfer will go away in the follow up PR after VS invoking this is tested

File: dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    auto packageFullNames{ AppModel::Package::FindByFamily(packageId.PackageFamilyName()) };
+    if (packageFullNames.empty())
+    {
+        wprintf(L"1 None found\n");
```
Comment (DrusTheAxe) [inline dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp:204]: Debugging printfer will go away in the follow up PR after VS invoking this is tested

File: dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        return true;
+    }
+    for (auto registeredPackageFullName : packageFullNames)
+    {
+        auto registeredPackage{ PackageId::FromPackageFullName(registeredPackageFullName) };
+        wprintf(L"C %ls %ls %d %d %I64X %I64X\n",
```
Comment (DrusTheAxe) [inline dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp:210]: Debugging printfer will go away in the follow up PR after VS invoking this is tested

File: dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+            continue;
+        }
+        if (registeredPackage.Version().Version >= packageId.Version().Version)
+        {
+            continue;
+        }
+
+        // This registered package is equal-or-better than the specified package.
+        // No need to register the specified package
+        wprintf(L"2 %ls not-needed because %ls\n", packageFullName, registeredPackage.PackageFullName().c_str());
```
Comment (DrusTheAxe) [inline dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp:225]: Debugging printfery will go away in the follow up PR after VS invoking this is tested

File: dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        return false;
+    }
+
+    // Nope. The specified package is better than anything current registered
+    wprintf(L"3 Nothing better\n");
```
Comment (DrusTheAxe) [inline dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp:230]: Debugging printfery will go away in the follow up PR after VS invoking this is tested

File: dev/Common/AppModel.Package.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+#ifndef __APPMODEL_PACKAGE_H
+#define __APPMODEL_PACKAGE_H
+
+#include <appmodel.h>
+
+namespace AppModel::Package
+{
+/// Find all Main+Framework packages in a package family registered to the current user
+inline std::vector<std::wstring> FindByFamily(PCWSTR packageFamilyName)
```
Comment (dhoehna) [inline dev/Common/AppModel.Package.h:12]: tab the method.

File: dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+int wmain(int argc, wchar_t *argv[])
+{
+    try
+    {
+        winrt::init_apartment();
+
+        // Parse the command line
+        int index{ 1 };
```
Comment (dhoehna) [inline dev/WindowsAppRuntime_MSIXInstallFromPath/main.cpp:251]: Is there any OSS tools that can be used to parse the command line?

**Pull request merged**